File: multiplication/src/main/java/learn/mssb/mssbmultiplication/challenge/ChallengeServiceImpl.java
package learn.mssb.mssbmultiplication.challenge;

import learn.mssb.mssbmultiplication.user.User;
import learn.mssb.mssbmultiplication.user.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Slf4j
@Service
@RequiredArgsConstructor
public class ChallengeServiceImpl implements ChallengeService {

    private final UserRepository userRepository;
    private final ChallengeAttemptRepository challengeAttemptRepository;

    @Override
    public ChallengeAttempt verifyAttempt(ChallengeAttemptDto attempt) {
        User user = userRepository.findByAlias(attempt.getUserAlias())
                .orElseGet(() -> {
                    log.info("Creating new user with alias {}", attempt.getUserAlias());
                    return userRepository.save(new User(null, attempt.getUserAlias()));
                });
        boolean isCorrect = attempt.getFactorA() * attempt.getFactorB() == attempt.getGuess();
        ChallengeAttempt challengeAttempt = new ChallengeAttempt(null, user,
                attempt.getFactorA(), attempt.getFactorB(),
                attempt.getGuess(), isCorrect);
        challengeAttemptRepository.save(challengeAttempt);
        return challengeAttempt;
    }

    @Override
    public List<ChallengeAttempt> getStatsForUser(String userAlias) {
        return challengeAttemptRepository.lastAttempts(userAlias);
    }
}

File: multiplicationkt/src/main/kotlin/challenge/ChallengeServiceImpl.kt
package learn.mssbkt.multiplicationkt.challenge

import learn.mssbkt.multiplicationkt.user.User
import learn.mssbkt.multiplicationkt.user.UserRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class ChallengeServiceImpl(
    private val userRepository: UserRepository,
    private val challengeAttemptRepository: ChallengeAttemptRepository
) : ChallengeService {
    private val log = LoggerFactory.getLogger(this::class.java)

    override fun verifyAttempt(attemptDTO: ChallengeAttemptDTO): ChallengeAttempt {
        val isCorrect = (attemptDTO.factorA * attemptDTO.factorB) == attemptDTO.guess
        val user: User = userRepository.findByAlias(attemptDTO.userAlias)
            .orElseGet {
                log.info("Creating new user with alias '{}'", attemptDTO.userAlias)
                userRepository.save(User(attemptDTO.userAlias))
            }
        val checkedAttempt = ChallengeAttempt(
            user = user,
            factorA = attemptDTO.factorA,
            factorB = attemptDTO.factorB,
            resultAttempt = attemptDTO.guess,
            correct = isCorrect,
        )
        val storedAttempt = challengeAttemptRepository.save(checkedAttempt)
        return storedAttempt
    }

    override fun getStatsForUser(alias: String): List<ChallengeAttempt> {
        return challengeAttemptRepository.lastAttempts(alias)
    }
}
